### PR TITLE
Allow a None batch dimension in SPSA for compatibility with generate_…

### PR DESCRIPTION
**Pull request by juesato**
…np()

Also add a test that the generate_np() interface works properly

Addresses https://github.com/tensorflow/cleverhans/issues/502

### CODE REVIEW

**Comment by catherio** [comment]
cc @nottombrown

File: cleverhans/attacks.py [hunk 1]
```diff
@@ -1821,3 +1821,8 @@ def loss_fn(x, label):
             is_debug=is_debug,
         )
         return adv_x
+
+    def generate_np(self, x_val, **kwargs):
+        # Add shape check for batch size=1, then call parent class generate_np
+        assert x_val.shape[0] == 1, 'x_val should be a batch of a single image'
```
Comment (nottombrown) [inline cleverhans/attacks.py]: Can we allow arbitrary batch sizes and then iterate through them in the function here? It seems like the "x_val must be a batch of a single image" constraint will probably be surprising to people.

File: cleverhans/attacks_tf.py [hunk 2]
```diff
@@ -1760,7 +1760,13 @@ def _get_delta(self, x, delta):
     def _compute_gradients(self, loss_fn, x, unused_optim_state):
         """Compute gradient estimates using SPSA."""
         # Assumes `x` is a list, containing a [1, H, W, C] image
```
[the remark this answers is not in the document]
Comment (nottombrown) [reply cleverhans/attacks_tf.py]: I agree with that logic. lgtm!

**Comment by resya173** [comment]
Hi

**Comment by npapernot** [comment]
The pep8 checker is failing: 

```
======================================================================
FAIL: Test if pep8 is respected.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/tensorflow/cleverhans/cleverhans/devtools/tests/test_format.py", line 45, in test_format_pep8
    raise AssertionError("PEP8 Format not respected")
AssertionError: PEP8 Format not respected
-------------------- >> begin captured stdout << ---------------------
/home/travis/build/tensorflow/cleverhans/cleverhans/attacks.py:1830:80: E501 line too long (80 > 79 characters)
--------------------- >> end captured stdout << ----------------------
```

**Comment by nottombrown** [comment]
Lgtm. Thanks @juesato !

**Pull request merged**